File: Projects/CPP-VNEXT/src/Plant.cpp
#include "Plant.h"
#include "World.h"
#include <random>

Plant::Plant(int strength, char symbol, Position&& initialPosition) : Organism(strength, 0, symbol, std::move(initialPosition)) {

}

void Plant::HandleAction(World& world) {
    static std::uniform_real_distribution<double> dist = std::uniform_real_distribution<double>(0, 1);
    static std::mt19937 rng = World::GetRng();

    if (dist(rng) < 0.65) {
        std::optional<Position> position = world.GetNearbyPosition(this->GetPosition(), 1, true);
        if (position) {
            world.Add(this->GetNewOfType(std::move(*position)));
        }
    }
}

void Plant::HandleCollision(CollisionContext& context) {}

File: Projects/Java/src/main/java/com/nieradko/worldsim/core/plants/Plant.java
package com.nieradko.worldsim.core.plants;

import com.nieradko.worldsim.core.*;

import java.util.concurrent.ThreadLocalRandom;

public abstract class Plant extends Organism {
    protected Plant(int strength, Position position) {
        super(strength, 0, position);
    }

    @Override
    protected void handleAction(IActionContext context) {
        if (ThreadLocalRandom.current().nextInt(100) < 10) {
            var proposedPosition = context.getRandomNearbyPosition(getPosition(), true);
            proposedPosition
                    .flatMap(this::getNewInstance)
                    .ifPresent(context::add);
        }
    }

    @Override
    protected void handleCollision(ICollisionContext collisionContext, IWorldContext worldContext) {}
}
